cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesStore.java

```diff
@@ -77,36 +83,46 @@ public PreferencesStore(DatasetFramework datasetFramework, TransactionSystemClie
     });
   }
 
-  private void setConfig(String namespace, String id,
+  private void setConfig(String namespace, EntityId entityId,
                          Map<String, String> propertyMap) throws NotFoundException, ProfileConflictException {
-    Config config = new Config(id, propertyMap);
-    validateAndSetPreferences(namespace, config);
+    Config config = new Config(entityId.toString(), propertyMap);
+    validateAndSetPreferences(namespace, entityId, config);
   }
 
-  private void deleteConfig(String namespace, String id) {
-    Transactionals.execute(transactional, context -> {
+  private void deleteConfig(String namespace, EntityId entityId) {
+    boolean profileExists = Transactionals.execute(transactional, context -> {
       try {
-        ConfigDataset.get(context, datasetFramework).delete(namespace, PREFERENCES_CONFIG_TYPE, id);
+        ConfigDataset configDataset = ConfigDataset.get(context, datasetFramework);
+        boolean exist = configDataset.get(namespace, PREFERENCES_CONFIG_TYPE, entityId.toString())
+                          .getProperties().containsKey(SystemArguments.PROFILE_NAME);
+        configDataset.delete(namespace, PREFERENCES_CONFIG_TYPE, entityId.toString());
+        return exist;
       } catch (ConfigNotFoundException e) {
-        //no-op
+        return false;
       }
     });
+
+    // remove the profile metadata since there is profile information in the preference
+    if (profileExists) {
+      profileMetadataPublisher.updateProfileMetadata(entityId);
```

Comment:
I put it outside since I don't want to break the transaction if the publish fails. I am not sure what is the correct way to deal with the publish failure